### ISSUE TITLE
Image/thumb format conversion

### ISSUE THREAD

**Issue by jenstornell** [k-next/plugins#54]
If no one else want to update this or make something similar, I will probably do it.

https://github.com/jenstornell/superthumbs

**Comment by bnomei** [comment]
a problem could be chaining the thumb/type variants generation between plugins.
my srcset plugin solves this by providing an option closure but limited to file, width and string of mime/type. 

https://github.com/bnomei/kirby3-srcset/blob/master/config.php

**Comment by bnomei** [comment]
@pedroborges seems to be working on this topic as well.
https://github.com/k-next/plugins/issues/19 -> 2nd post

**Comment by pedroborges** [comment]
Mine extends Kirby thumbnail generator. It let's Kirby resize/crop as requested but setting quality at 100, after Kirby is done it passes the image through a chain of optimizers (as decided by the developer via a configuration). There are optimizers for jpeg, png, and gif. If webp is added to the optimizers chain it will make a copy of the image in that format too.

I still need to add an option to let you define presets and also add some file methods to help you choose which chain gets used. I'll be testing it two projects in August and release when it's more robust. Right now it's working really well and the compression results are great!

**Comment by bnomei** [comment]
imagekit used to kind of jobs logic to allow pregeneration and delaying of thumb creation. why is that a nice idea?

since some of my website have 300 jpg images on homepage (lazyloaded in galleries) with about 5 srcset variants. so if i would use your plugin and enable the webp variant on first load the page would try to create and optimize `300 x 5 x 2 = 3000` images in on call. even if refreshed multiple times this might take about about 30 minutes on average hosting setups. imagekit solves this by delaying the srcsets and variants until requested.

so maybe you can find a nice solution for that as well?

**Comment by pedroborges** [comment]
@bnomei I don't need to worry about that in Kirby 3 because that's the new behaviour. Kirby creates a job for each thumbnail URL and they are generated individually as images are requested.

In your example, the page would load fast and the images would be created as they are lazy loaded. So only the srcset variant picked by the browser would be created for each user.

**Comment by jenstornell** [comment]
I would love some kind of chained approach. Hang on compression, or image format conversion, crop or all of it.

**Comment by bnomei** [comment]
maybe with jens input we can focus on just pedros plugin? so closing this issue as discussion?

**Comment by bnomei** [comment]
@pedroborges will you create a new issue explaing what your plugin does and how the creation is triggerd? so i could check how to link/chain it to my srcset plugin?

**Comment by jenstornell** [comment]
If his plugin is as simple as my old one and works well, I can close the issue.

I just want to get there first. I'll be happy to close my plugin if a better one arrives. ☺

**Comment by bnomei** [comment]
@jenstornell i agree with you jens. it seems like pedros plugin might do a lot. maybe he could make all the converter libs/bin optional downloads?

**Comment by bnomei** [comment]
@pedroborges @jenstornell m srcset plugin allows registering types and a custom resize function. do you think this will work well with your ideas of the image type/format converter plugin?

it iterates over all types and echos `picture > source` elements.

https://github.com/bnomei/kirby3-srcset/#options-explained
https://developer.mozilla.org/en-US/docs/Web/HTML/Element/source
https://github.com/k-next/kirby/blob/9c20612208655c55f811fd105d79f57123e3abe2/src/Toolkit/Mime.php

so to create source for jpg and webp one would set in config...

```php
'types' => ['image/jpeg', 'image/png', 'image/webp'], // MIME
'resize' => function ($file, $width, $type) {
    // NOTE: override and do something with $type
    // like: $file->convert($width, $type);
    return $file->resize($width);
}
```

**Comment by jenstornell** [comment]
@bnomei I don't really answer your question but I think something like this would be nice.

```php
$page->image('image.jpg')->crop(100, 200)->convert('jpg', 80)->compress();
```

**This is what it could do:**

1. Crop to 100x200, like it works in the core now.
1. If it's a `png` or other pixel format, convert to `jpg`. If already `jpg`, do nothing in this step. It also uses quality 80.
1. Even if the quality is 80, the image is not compressed. It does that here.

**Another example could be:**

```php
$page->image('image.png')->resize(100)->compress();
```

It resizes an image and then compress it.

**Comment by bnomei** [comment]
@jenstornell
- in `->convert('jpg', 80)` the `jpg` should be a mime type not a file extension? and if its an file extension the converter plugin should accept mime types as well.
- the return value of `compress()` must be a  proper`Kirby\Cms\File` object not just a string to root (which probably is wha you suggested, but just to make sure).

**Comment by jenstornell** [comment]
@bnomei Both your additions sounds good to me. In my Superthumbs plugin I only had `jpg`, but I agree that mime is more correct, but for simplicity `jpg` could be a shorthand for it.

**Comment by jenstornell** [comment]
I just found this: https://github.com/spatie/image-optimizer

I even tried to make it a plugin but for some reason adding a thumb component with a plugin did not work.

```php
Kirby::plugin('your/plugin', [
  'components' => [
    'thumb' => function (App $kirby, string $src, string $dst, array $options) {
      echo 'test';
      die;
    }
  ]
]);
```

To make sure it was called I did this:

```php
$screenshot = $page->images()->filterBy('name', '*=', 'screenshot')->first();
$thumbnail = $screenshot->crop(900, 450, 'top');
```

But nope.

**Comment by bnomei** [comment]
@jenstornell not sure exactly if echoing will do anything with new jobs logic. creating files should. check out my thumbs component for imageopzim to get started.

**Comment by bnomei** [comment]
sry. i forgot to post the link: https://github.com/bnomei/kirby3-thumb-imageoptim/blob/master/classes/imageoptim.php

**Comment by bnomei** [comment]
if you want to convert image to another type you can use:
https://github.com/bnomei/kirby3-cloudconvert
